### ISSUE TITLE
SSH connections always look to agent for authentication even when key_file is provided

### ISSUE THREAD

**Issue by sathiyas** [hashicorp/terraform#2614]
The below used to work before till 0.5.3 without needing specify host within connection. Now SSH connection does not work as waits for ever and timesout. Seems it does not have correct IP. Tried host = "${self.private_ip}", no luck, any clues? 

```js
resource "aws_instance" "server" {
    ami = "${lookup(var.ami, concat(var.region, "-", var.platform))}"
    instance_type = "${var.instance_type}"
    count = "${var.servers}"

  connection {
        user = "${lookup(var.user, var.platform)}"
        #host = "${self.private_ip}"
        key_file = "${var.key_path}"
    }
}
```

**Comment by phinze** [comment]
Hi @sathiyas - I assume the ssh connection that does not work is via a `remote-exec` provisioner?

Testing locally to verify over here.

**Comment by sathiyas** [comment]
Yes, this is the code, it fails at file provisioner right away

resource "aws_instance" "server" {
    ami = "${lookup(var.ami, concat(var.region, "-", var.platform))}"
    instance_type = "${var.instance_type}"
    key_name = "${var.key_name}"
    count = "${var.servers}"
    security_groups = ["${var.consul_sg}"]
    subnet_id = "${var.subnet_id}"
    connection {
        user = "${lookup(var.user, var.platform)}"
        #host = "${self.private_ip}"
        key_file = "${var.key_path}"
    }

    #Instance tags
    tags {
        Name = "${var.tagName}-${count.index}"
        Name = "${var.ASV}"
    }

    provisioner "file" {
        source = "${path.module}/scripts/${var.platform}/upstart.conf"
        destination = "/tmp/upstart.conf"
    }

    provisioner "file" {
        source = "${path.module}/scripts/${var.platform}/upstart-join.conf"
        destination = "/tmp/upstart-join.conf"
    }

    provisioner "remote-exec" {
        inline = [
            "echo ${var.servers} > /tmp/consul-server-count",
            "echo ${aws_instance.server.0.private_dns} > /tmp/consul-server-addr",
        ]
    }

    provisioner "remote-exec" {
        scripts = [
            "${path.module}/scripts/${var.platform}/install.sh",
            "${path.module}/scripts/${var.platform}/server.sh",
            "${path.module}/scripts/${var.platform}/service.sh",
        ]
    }
}

**Comment by phinze** [comment]
@sathiyas still testing over here - can you add `agent = false` to your connection block and see if that changes the behavior at all?

**Comment by sathiyas** [comment]
Cool, it worked with agent=false, I should have tried that :( 

I will submit couple of PRs to add this to existing examples wherever connection is used

**Comment by phinze** [comment]
@sathiyas you shouldn't need `agent = false` for it to work. I found the bug causing this - need to reorder the list of available authentication methods provided to SSH.

Tagging as a core bug and will get this fixed!

**Comment by sathiyas** [comment]
great, thx

**Comment by ajlanghorn** [comment]
I've just experienced this as well. A quick hunt on the mailing list threw up a link to this issue. Adding `agent = false` for the meantime, as that works. Looking forward to 0.6.1. Thanks @phinze!

**Comment by romanvbabenko** [comment]
still got it
Terraform 0.6.1 agent=false doesn't work for me. Digital Ocean provider cannot connect via remote-exec.

**Comment by ajlanghorn** [comment]
I concur re: DO. My current project is all in EC2, and that appears to work okay. @romanvbabenko - do you get anything interesting from TF_LOG?

**Comment by phinze** [comment]
Hi @romanvbabenko and @ajlanghorn - can you open a fresh issue for the digitalocean problem with an example config? Thanks!

**Comment by tonymurphy** [comment]
had a similar problem, turns out my ssh username was incorrect..

**Comment by EntilZha** [comment]
Ran into same issue as @tonymurphy on AWS, is there a way to set that globally or does it have to be per connection block?

**Comment by Vishal-Gaur** [comment]
Hi All,

I'm also facing this issue, please suggest.

root@gfs1:/etc/heketi# heketi-cli topology load --json=/etc/heketi/topology.json

	Found node gfs1 on cluster 8f207c20e4470125f11b3a98306c98a8
		Adding device /dev/vda1 ... Unable to add device: ssh: handshake failed: ssh: unable to authenticate, attempted methods [none publickey], no supported methods remain
	Creating node gfs2 ... Unable to create node: ssh: handshake failed: ssh: unable to authenticate, attempted methods [none publickey], no supported methods remain

Thanks

**Comment by danistrate** [comment]
Hello Nathan, 
I have the same issue as yourself, but with no workaround so far. The weird thing is that I can easily SSH into an AWS instance and use remote-exec to run some commands, but when trying the same on the Oracle Public Cloud I get this error.

**Comment by nathanhinchey** [comment]
@Vishal-Gaur @danistrate I clarified my comment with my workaround, hopefully that helps you.

**Comment by danistrate** [comment]
Did it for me, thanks!

**Comment by pmoosh** [comment]
I think there is a regression for 0.11.4. I didn't have this issue 0.11.3....

**Comment by nathanhinchey** [comment]
For me the issue was that I added a new RSA key not named `id_rsa` and
until I used `ssh-add` it gave this error. I don't think this is actually
the same issue as #2614 -- sorry for replying to this thread, it was just
what I got to when googling the error message and I wanted to provide my
workaround.

On Mar 23, 2018 21:45, "Peter Mooshammer" <notifications@github.com> wrote:

I think there is a regression for 0.11.4. I didn't have this issue
0.11.3....

—
You are receiving this because you commented.
Reply to this email directly, view it on GitHub
<https://github.com/hashicorp/terraform/issues/2614#issuecomment-375837712>,
or mute the thread
